### DISCUSSION TITLE
About adding new buttons and tweaking behavior

### DISCUSSION THREAD

**Discussion by lune36400** [bazmonk/digigurdy-baz#32] [General]
today i try to upload on teensy 3.5 but fatal error
![image](https://user-images.githubusercontent.com/68997708/157472245-28d5d358-f9b7-4cbc-ad5b-8611971a4ee3.png)

the error is 

Arduino : 1.8.13 (Windows 10), TD: 1.56, Carte : "Teensy 3.5, Serial + MIDI, 120 MHz, Faster, French"

C:\Users\Hervé\AppData\Local\Temp\Temp1_digigurdy-baz-main (2).zip\digigurdy-baz-main\digigurdy-baz\digigurdy-baz.ino:123:46: fatal error: bitmaps.h: No such file or directory

compilation terminated.

Erreur de compilation pour la carte Teensy 3.5

Ce rapport pourrait être plus détaillé avec
l'option "Afficher les résultats détaillés de la compilation"
activée dans Fichier -> Préférences.

**Comment by bazmonk** [comment]
There are three (as of the latest version late last night... your version may not have all of them) extra header files that need to be in the same directory/folder as the `digigurdy-baz.ino` file when you compile/verify/upload:

```
// These are found in the digigurdy-baz repository
#include "bitmaps.h"       // Pretty pictures
#include "eeprom_values.h" // Save-slot memory addresses
#include "default_tunings.h" // Preset tunings.
```

Can you check those files are present?  It looks like they're just missing.

**Comment by lune36400** [comment]
sorry it's ok now,

I think this night i need for my new pcb, because is a combo  dg et hg

no change for the menu and detect crank motor, the soft is ok for me,

-at least,  the dg run directly on play (without push gaming button, stay like that if i want play dg only.

-a button to stop open string (the drone can be set in menu or in bs16i)

-a button to stop when hg play without dg all the time (if all the time i cut dg)
 - the 3 button for capo
 -2 button for  x and o

i think i must stop the pcb before all the new soft is design.

i can add and comment and uncomment if you telle me where

(where you come from, fb is unpersonnal, me to center of france near st chartier (festival 3km)

in the future possibility to play in dg more than 1 note in the same time, to do "accord"

thanks a lot

rv

![image](https://user-images.githubusercontent.com/68997708/157486676-a02fb5b4-1bf8-408d-87bc-c0eed438d79f.png)

**Comment by bazmonk** [comment]
I'm not sure if I understand all the needs/questions correctly, but let me show you a few parts of the code and it might help you figure out what you'd need to change to support duplicated buttons or different string behavior.

First, however, late last night I added in a control in the pause menu to turn the drone and trompette on and off, and you can also turn them individually on and off.  So that control is also present now.  I can't tell but I think it might have some bearing on how you work this.

In the code, I make different button objects to control interacting with them in the logic of the code.  I have a `GurdyButton` object which is a generic push button like the capo button.  I have a `KeyboxButton` which is an extension of that and has a note associated with it: they manage the musical keys.  And finally I have a `ToggleButton` which I wrote for the big gaming button.  It has an added sense of being toggled on and off instead of just press=on, release=off.

To add in new physical buttons, you just need to create an object for them.  Here's where I declare all the buttons and objects in the code.  I do this outside of any of the functions near line 700:

```
// Declare the "keybox" and buttons.
HurdyGurdy *mygurdy;
ToggleButton *bigbutton;
GurdyCrank *mycrank;

// As musical keys, these are referred to in the mygurdy object above.
// This declaration of them is specifically for their use as navigational
// buttons in the menu screens.  ok = O, back = X.
KeyboxButton *myOkButton;
KeyboxButton *myBackButton;
KeyboxButton *my1Button;
KeyboxButton *my2Button;
KeyboxButton *my3Button;
KeyboxButton *my4Button;
KeyboxButton *my5Button;
KeyboxButton *my6Button;

// Note that there aren't special classes for melody, drone, even the keyclick.
// They are differentiated in the main loop():
// * A melody string is one that changes with the keybox offset.
// * A drone/trompette is one that doesn't change.
// * The keyclick "string" is just a drone that comes on and off at particular times.
// * The buzz "string" is also just a drone that comes on/off at other particular times.
GurdyString *mystring;
GurdyString *mylowstring;
GurdyString *mykeyclick;
GurdyString *mytromp;
GurdyString *mydrone;
GurdyString *mybuzz;

GurdyButton *tpose_up;
GurdyButton *tpose_down;
GurdyButton *capo;
```

The `pin_array` at the beginning of the code that determines what the keybox will be is used by the HurdyGurdy object and it handles creating the keybox buttons.  That's why you don't see all 24 of them here.

After that, I initialize the button objects, and that's where they get their pin assignments.  This happens inside setup():

```
// The keybox arrangement is decided by pin_array, which is up in the CONFIG SECTION
  // of this file.  Make adjustments there.
  mygurdy = new HurdyGurdy(pin_array, num_keys);
  bigbutton = new ToggleButton(39);

  // Grab the keys we use for menu nav out of the keybox:
  myBackButton = mygurdy->keybox[0];
  myOkButton = mygurdy->keybox[num_keys - 2];

  // 1-6 keys are the first six bottom keys of the keybox
  my1Button = mygurdy->keybox[1];
  my2Button = mygurdy->keybox[3];
  my3Button = mygurdy->keybox[4];
  my4Button = mygurdy->keybox[6];
  my5Button = mygurdy->keybox[8];
  my6Button = mygurdy->keybox[9];

  tpose_up = new GurdyButton(21);   // A.k.a. the button formerly known as octave-up
  tpose_down = new GurdyButton(22); // A.k.a. the button formerly known as octave-down

  capo = new GurdyButton(23); // The capo button
```

You'd want your new buttons to be defined here.

Then, inside `loop()`, all the logic happens.  It begins with running the `update()` method to poll the buttons:

```
// Update the keys, buttons, and crank status (which includes the buzz knob)
  myoffset = mygurdy->getMaxOffset();  // This covers the keybox buttons.
  bigbutton->update();
  mycrank->update();
  tpose_up->update();
  tpose_down->update();
  capo->update();

  myOkButton->update();
  myBackButton->update();
```

And then just a bunch of if this else that statements about what to do if certain combinations are pressed.  For example, this is the part that checks if the X and O buttons were pressed.  If they were, I'm turning off the sound on all the strings, I'm toggling off the gaming button (so it doesn't come back on afterwards by itself), and then I'm running the function that goes through the pause screen options:

```
// If the "X" and "O" buttons are both down, trigger the tuning menu
  if (myOkButton->beingPressed() && myBackButton->beingPressed()) {

    // Turn off the sound :-)
    mystring->soundOff();
    mylowstring->soundOff();
    mykeyclick->soundOff();  // Not sure if this is necessary... but it feels right.
    mytromp->soundOff();
    mydrone->soundOff();
    mybuzz->soundOff();

    // If I don't do this, it comes on afterwards.
    bigbutton->setToggle(false);

    pause_screen();
```

There's an `if` that looks for keybox keys changing, for the transpose buttons, for the buzz/crank, for the gaming button... lots of examples for you to see either where to insert your button actions, or how to customize what it's doing.

Does that help at all?
